### PR TITLE
Adjust to numpy dev scalar format

### LINKED ISSUE #15095: TST: 400+ doctest failures because numpy 2.0.dev changed string representation
Example log:

https://github.com/astropy/astropy/actions/runs/5671453922/job/15368606312

Too many to post here but this might give you an idea:

```
Numpy: 2.0.0.dev0+629.g511f77d12
...
      at [0, 2]:
  -     a> 2
  -     b> 3
  +     a> np.int64(2)
  +      ?          ^
  +     b> np.int64(3)
  +      ?          ^
      ...and at 6 more indices.
```

Caused by:

* https://github.com/numpy/numpy/pull/22449

Also related maybe:

* #14553

I think @mhvk has a cunning plan? 🙏

### LINKED ISSUE #16491: Reminder: clean up `without_legacy_printoptions` test decorator
This is an issue to remind ourselves to remove `@without_legacy_printoptions` everywhere it's used as well as its definition when it's not needed anymore

What should happen first (ideally, in this order):
- #16423
- #15065 

For context I don't expect this issue to ever become urgent: this decorator will become a noop when #15065 is merged, but that's also the reason why it should be removed in the first place :)

### PR DESCRIPTION
**Pull request by mhvk**
DEV NOTE: Do not merge until numpy 2.0 is actually out! ~Do not merge until this is merged and available in numpy nightly wheel (when we start to see things fail in devdeps)~

- [x] Revert https://github.com/astropy/astropy/pull/15096 if it got in first. (EDIT: it should get in first; this one is for when 2.0 is actually out)
- [x] Decide for scalar quantity.value whether it should not just return an array scalar (see below) -> made this its own issue: #16508
- [x] Check whether we may want similar adjustments elsewhere (maybe as follow-up; e.g., `Structure` for structure units output) - but this should not be part of this PR, so punted to #15793.
- [x] Decide whether we are happy for the docs to pass tests only on numpy >=2.0. If not, we have to update `doctest-plus` to ignore `np.float64`, etc., in comparing numerical values. 2023-Dec: seems that we are heading to a docs-on-numpy>=2.0 only state

<hr/>

This is a start to adjusting our tests for the change in how scalars are formatted in NEP 51 - see https://github.com/numpy/numpy/pull/22449

~Note that it does not adjust tests of `repr` and the documentation yet!~

Posting this in part to show that outside of `repr` this is not a big deal.

In the documentation, one often will have things like
```
>>> q = 5 * u.m
>>> q.value
5.0  -> np.float64(5.0)
```
This brings up a more general question of whether we should use this opportunity of the `repr` changing to just return an array scalar (i.e., it would be `np.array(5.0)` instead). EDIT: this is better discussed elsewhere - see #16508.

Fix #15095
Fix #16491

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
The diff looks much simpler than I anticipated. Thanks! 🤞

**Comment by mhvk** [comment]
Well, the main diff from the docs is not yet in there...

**Comment by pllim** [comment]
If https://github.com/astropy/astropy/pull/15096 goes in first, that PR also has to be reverted here. FYI.

**Comment by mhvk** [comment]
I just rebased and force-pushed to get a fresh test run. If `-dev` passes, we probably should run `extra CI` as well, just in case.

p.s. Very pressed for time right now, so if this doesn't converge easily, happy with #15096 as a stop-gap measure...

**Comment by pllim** [comment]
Alas... still 95 failed (though 3 of them are from modeling and unrelated).

**Comment by pllim** [comment]
Looks like we need version check in some places?

`ImportError: cannot import name 'get_formatter' from 'numpy.core.arrayprint'`

**Comment by mhvk** [comment]
Ah, right, that changed, we didn't make the new `get_formatter` in the end!  Can be a quick change (hopefully)...

**Comment by pllim** [comment]
There are 51 failures now (3 unrelated, the rest doctest output comparison failures). I actually don't like the new look in our own doc test output... especially when you print out the WCS... it looks jarring and adds noise to the actual data I am interested in (say, what is the value of CVAL1, without also have to see it is `np.float` because I already know that).

**Comment by mhvk** [comment]
The problem with this PR is that all doctests will fail on older numpy. In principle, we can just not run the tests on older versions, but that makes sense only once 2.0 is actually out. I think the logical route is to go with #15096 first, so that things work again, and then do this one once 2.0 is released.

File: astropy/units/structured.py
Comment (pllim) [inline astropy/units/structured.py:124]: How does @namurphy and @nstarman feel about this?

**Comment by saimn** [comment]
@mhvk - Do we need this PR in Astropy 6.0 or can we postpone ?

**Comment by mhvk** [comment]
That's a bit tricky, especially as we have not quite decided how to do it with the documentation - with this merged, the documentation runs will fail on older numpy. In principle, not a big problem, as arguably it makes sense to test the documentation only on the most recent versions of packages we support. Alternatively, we'd have to adjust our doctest machinery to systematically ignore `np.float64(...)`, etc. (that may not be that hard).

But the argument that our documentation should be consistent with the most recent version of dependencies also means this should probably be postponed, since numpy 2.0 seems likely to appear after astropy 6.0. So, let me change the milestone...

**Comment by mhvk** [comment]
p.s. I rebased just to have things up to date.

**Comment by pllim** [comment]
Thanks! But I think we need to postpone this because it makes no sense to run the stable doc test on unreleased numpy.

**Comment by mhvk** [comment]
This disables the doctests in `.rst` files and some python files with problematic docstrings for numpy < 2, so that in principle all the CI should pass again. Still only for when numpy 2.0 is actually out.

File: astropy/stats/jackknife.py
Comment (astrofrog) [inline astropy/stats/jackknife.py:121]: I'm confused, I thought a few years ago we were told numpy's float64 was an alias for the pure Python float? In any case I wonder if we should consider whether what we actually want to do is return numpy scalars or Python scalars now that they look different?

Comment (neutrinoceros) [reply astropy/stats/jackknife.py:121]: > I thought a few years ago we were told numpy's float64 was an alias for the pure Python float?

@astrofrog maybe you got this from Numpy 1.20's release notes about deprecated type alias ?
https://numpy.org/doc/stable/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated

Also note that [Python's doc](https://docs.python.org/3/tutorial/floatingpoint.html#representation-error) mentions
> and almost all platforms map Python floats to IEEE 754 binary64 “double precision” values

so numpy's `float64` cannot be aliased to `float` on all platforms.

Comment (mhvk) [reply astropy/stats/jackknife.py:121]: It is indeed not the same - it has `.dtype`, etc. Indeed, this is why the change was made - to make clear to the user that they are not identical -- which of course is more useful for single precision, different integers, etc.

We could indeed reconsider what we return, though I think in the end it does not matter much.

**Review by taldcroft — COMMENTED** [review]
Mostly OK for io.ascii, time, table. Thanks, this looks like  a lot of work!!

File: astropy/table/row.py
Comment (mhvk) [reply astropy/table/row.py:133]: Numpy 2.0 also changes the default integer for windows...

Comment (taldcroft) [reply astropy/table/row.py:133]: Hallelujah!

File: astropy/units/structured.py
Comment (nstarman) [reply astropy/units/structured.py:126]: Agreed! I think the current ``((PhysicalType('length'), PhysicalType({'speed', 'velocity'})), PhysicalType('time'))`` should be the behaviour of `__str__`.
For `__repr__` we at least don't want to show the module `structured`.

```suggestion
        astropy.units.Structure((astropy.units.Structure((PhysicalType('length'), PhysicalType({'speed', 'velocity'})), dtype=[('f0', 'O'), ('f1', 'O')]), PhysicalType('time')), dtype=[('f0', 'O'), ('f1', 'O')])
```

Comment (mhvk) [reply astropy/units/structured.py:126]: Agreed, that is the minimal change. But coming back to this fairly substantial PR, it is probably best to do that afterwards - I raised a new issue at #15793.

Comment (nstarman) [reply astropy/units/structured.py:126]: Totally fine by me. I can review this PR in a few days, if you're ready @mhvk

Comment (mhvk) [reply astropy/units/structured.py:126]: Thanks! This PR is basically waiting for numpy 2.0 to be released...  I'm just keeping it up to date once in a while until that happens.

**Comment by neutrinoceros** [comment]
When do we need to merge this ? Could it be included in 6.1.1 or does it need to wait on numpy 2.0.0 final ?
@pllim @mhvk @astrofrog @taldcroft

**Comment by taldcroft** [comment]
> Our CI mostly grabs numpy stable, so I don't think we can do this before numpy 2.0 stable is released?

So astropy is broken for numpy 2.0 but we can't merge this fix until 2.0 is released? Doesn't the `CI / allowed_failures / (Allowed Failure) Python 3.12 with remote data and dev version of key dependencies (pull_request)` test get numpy-dev which is 2.0.x?

Does anybody understand the first sentence of this PR `DEV NOTE: Do not merge until numpy 2.0 is actually out!`? Was that just about CI testing or something deeper?

**Comment by pllim** [comment]
The fix for https://github.com/astropy/astropy/issues/15792 should be backported by not https://github.com/astropy/astropy/issues/15095 . Are the fixes for each separable?

**Comment by taldcroft** [comment]
Ping @mhvk - this needs a rebase for merge conflicts and there is some urgency with the looming numpy 2.0 final release. 

Personally I think that numpy 2.0-only doc testing is fine (for the fourth bullet in the description).

BTW, do we have a better estimate of the release at this point?

**Comment by pllim** [comment]
> numpy 2.0-only doc testing is fine

It is fine but if done too early, we basically are skipping the doctest until numpy 2.0 is released.

**Comment by pllim** [comment]
Re: numpy 2.0 release -- Should have happened by now according to https://github.com/numpy/numpy/issues/24300 🤷‍♀️

**Comment by neutrinoceros** [comment]
> BTW, do we have a better estimate of the release at this point?

I think numpy devs are waiting for a wider adoption in the ecosystem so the final release breaks as as little projects as possible, see https://github.com/numpy/numpy/issues/26191
but the exact target is still unspecified, as far as I know

**Comment by neutrinoceros** [comment]
This is also my expectation looking the diff for `yaml.py`, which is mostly changing `!r` for `!s` (e.g., exactly what I was looking to do earlier today when chasing the duplicate issue #16414)

**Comment by mhvk** [comment]
I think you all converged already, but as @pllim was saying, we were hoping to ensure the doc tests would always run, so were waiting for 2.0 final for that reason. Please do feel free to cherrypick commits! I tried to keep them separate exactly for that.

**Comment by mhvk** [comment]
While testing this with `test-devdeps-alldeps`, I found some tests with pandas failed. Since these are not representation specific, I'm ignoring them here...  Also, if `bottleneck` is installed `stats.sigma_clipped_stats` uses `float` not `np.float64` so the `repr` looks different. Not quite sure how to deal with that.

**Comment by pllim** [comment]
For `stats` + `bottleneck`, maybe @larrybradley has opinions. 🤞

**Comment by larrybradley** [comment]
My best guess is that we should make them consistent if possible.  Is there any easy way to do that?

**Comment by mhvk** [comment]
> My best guess is that we should make them consistent if possible. Is there any easy way to do that?

Yes, it is not very difficult: either turn the regular non-bottleneck output into `float` (if it is not an array), or the bottleneck output into `np.float64`. The latter may be slightly preferable, since it is easier to deal with a numpy scalar in places where one might also get an array (if `axis` is passed in).

Note that this difference almost certainly has been around from the get-go -- the only thing that changed in numpy 2.0 is the `repr`, which for `np.float64` now is different from that of `float` -- this is what caused a doctest to fail.

**Comment by larrybradley** [comment]
@mhvk Thanks I'll open a PR to change the bottleneck output to `np.float64`.

**Comment by mhvk** [comment]
> While testing this with `test-devdeps-alldeps`, I found some tests with pandas failed.

See fix in #16439. Combined with @larrybradley's fixes in #16431 that I already rebased on, everything should pass, which will mean this can go out of draft. Note that #16439 will bring in a `devdeps-alldeps` run in extra CI that would be good for a final sanity check.

File: astropy/io/votable/tests/test_vo.py
Comment (neutrinoceros) [inline astropy/io/votable/tests/test_vo.py:270]: I'm don't think this is the correct fix here. I'll link to my own solution when I open the PR.

Comment (neutrinoceros) [reply astropy/io/votable/tests/test_vo.py:270]: https://github.com/astropy/astropy/pull/16442

**Comment by neutrinoceros** [comment]
All non-doctests parts of this PR were sliced out and backported. We should now be able to rebase this branch in preparation for the upcoming numpy 2.0.0 final (June 16th). I think the sooner this is merged the better, because new related bugs might slip in in the mean time.

**Comment by mhvk** [comment]
@neutrinoceros - good point!  I rebased, minimized the new additions to `pyproject.toml`, and removed the `printoptions` stuff.

**Comment by neutrinoceros** [comment]
great ! This now closes #16491

File: astropy/table/table_helpers.py
Comment (neutrinoceros) [inline astropy/table/table_helpers.py:108]: Can you confirm this function looks is *only* meant for testing and simple doc examples ?

Comment (mhvk) [reply astropy/table/table_helpers.py:108]: Yes, confirmed!

**Comment by mhvk** [comment]
@mcara, @larrybradley, @modeling, @saimn, @cosmology, @WilliamJamieson, @matteobachetti - this PR changes docstrings and documentation to use numpy 2.0 `repr`. It involves no code changes, but does mean the documentation tests now are only done on numpy >= 2.0 (unavoidable). If you want to have a simpler look, check commits specifically to your subpackage. Though really I think one person checking this should be good enough.

**Comment by mhvk** [comment]
Note that I removed the backport label - I think for 6.1, we should keep the doctests on numpy 1.26, but for 7.0 we switch to 2.0.

**Review by neutrinoceros — APPROVED** [review]
FWIW I already had a look at everything left in the PR and it looks good to me. I also agree this is best not backported, unless it turns out to create friction after numpy 2.0 is release, which as far as I understand shouldn't be the case.

**Comment by pllim** [comment]
Just so we're on the same page, are we still waiting for numpy 2.0 stable to come out before merging?

File: pyproject.toml
Comment (mhvk) [reply pyproject.toml:215]: OK, maybe that's an argument to change the requirements - I don't really want to have to think about this again...

Comment (pllim) [reply pyproject.toml:215]: So, is your proposal to change it to `numpy>=2.0.dev` and then we temporarily make devdeps passing required for merging?

Comment (mhvk) [reply pyproject.toml]: OK, I changed the requirement to 2.0.0rc1, so we get a real failure if something changes.

**Comment by mhvk** [comment]
Hmm, that's annoying, the fix for `boost` didn't actually work - see failure in daily cron. Turns out #16431 had not caught all occurrences...

@larrybradley - I redid your fix from #16431 in a different way, which makes everything pass. Could you check the last commit, 840f022389b5cc5c521d835d4936e38e5bbb9b75?

**Comment by mhvk** [comment]
@pllim - the pre-release test in daily cron now indeed includes the doc tests (and passes), so we will get warnings in the unlikely case there is a change to the documentation between now and the official release.

**Comment by larrybradley** [comment]
@mhvk For `sigma_clip`, `cenfunc` and `stdfunc` can also be input as user-defined callable functions.  Do we care if such functions return Python floats?  That's why in #16431 I converted all return values from `cenfunc` and `stdfunc` into np.float.

**Comment by mhvk** [comment]
@larrybradley - ah, I see. At some level, no, we do not care - ducktyping is good python practice, after all.  And really, even here the only reason we found this is because the doctests cannot both pass with and without bottleneck otherwise. Though of course it is nice if the types astropy itself produces do not depend on what package one happens to have installed!

Anyway, I think I'd leave it as it is now...

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/15065#issuecomment-2135791918

Does this mean the Daily Cron job has to be run now for every PR until numpy 2 is out? 🤔

**Comment by mhvk** [comment]
> Re: [#15065 (comment)](https://github.com/astropy/astropy/pull/15065#issuecomment-2135791918)
> 
> Does this mean the Daily Cron job has to be run now for every PR until numpy 2 is out? 🤔

That would be one solution, but so few PRs touch the documentation that I think we could just do without...

Obviously, another option is still just to wait for the actual release.

**Comment by pllim** [comment]
We could temporarily add a predeps job to the regular CI as part of this PR, and then make a reminder issue to remove it after. I think we are still a good 3-4 weeks until numpy 2 is released. Not sure if we need remote data though...

https://github.com/astropy/astropy/blob/7202529f032fadf68d8757614ab87979d8096719/.github/workflows/ci_cron_daily.yml#L48-L53

**Comment by mhvk** [comment]
How about I change the double run to use `predeps`? Maybe not even temporarily: it seems somewhat sensible in general to have a predeps run on our important dependencies.

**Comment by pllim** [comment]
Double run uses the test runner, it seems. I remember the test runner not very good with doctest (https://github.com/astropy/astropy/issues/8153) so not a good candidate.

https://github.com/astropy/astropy/blob/6f217f4923df660884c681a08b1c02542be29763/tox.ini#L120

**Comment by mhvk** [comment]
OK, I stuck it on the first run with python 3.12 and minimal dependencies.

**Comment by mhvk** [comment]
Removed Extra CI since we just checked that worked.

File: .github/workflows/ci_workflows.yml
Comment (pllim) [inline .github/workflows/ci_workflows.yml:68]: Maybe this is okay for a few weeks. It is good to have doctest in coverage.

**Review by pllim — APPROVED** [review]
Thanks! My goal is to retire before Numpy 3 or Python 4 comes out, whichever first.

**Comment by pllim** [comment]
Thank you so much for going through this pain, @mhvk ! I opened https://github.com/astropy/astropy/issues/16512 as a follow-up.

**Comment by mhvk** [comment]
Thanks!!

**Pull request merged**
